### ISSUE TITLE
The Azure Storage APIs are a bit confusing and misleading

### ISSUE THREAD

**Issue by IEvangelist** [dotnet/aspire#6875]
## Background and Motivation

The current Azure Storage API is a bit confusing and misleading. Today, when you call `AddAzureStorage` it adds provisioning and infrastructure details, including roles, to the Azure storage resource for blobs, queues, and tables. This happens regardless of whether or not you call `AddBlobs`, `AddQueues`, or `AddTables`. It might make more sense to instead change these specific APIs to be `GetBlobs`, `GetQueues`, and `GetTables` instead without the name parameter.

## Proposed API

```csharp
namespace Aspire.Hosting;

public static class AzureStorageExtensions
{
    public static IResourceBuilder<AzureStorageResource> AddAzureStorage(
        this IDistributedApplicationBuilder builder, [ResourceName] string name)
    {
        // ...
    }

   // Omitted for brevity...

    public static IResourceBuilder<AzureBlobStorageResource> GetBlobs(
        this IResourceBuilder<AzureStorageResource> builder);

    public static IResourceBuilder<AzureTableStorageResource> GetTables(
        this IResourceBuilder<AzureStorageResource> builder);

    public static IResourceBuilder<AzureQueueStorageResource> GetQueues(
        this IResourceBuilder<AzureStorageResource> builder);
}
```

## Usage Examples

```csharp
var builder = DistributedApplication.CreateBuilder(args);

var storage = builder.AddAzureStorage("storage").RunAsEmulator(static container =>
{
    container.WithDataBindMount();
});

var blobs = storage.GetBlobs();
var queues = storage.GetQueues();
var table = storage.GetTables();

// Use storage bits...

builder.Build().Run();
```

## Risks

This is a breaking change, and will require obsoleting existing `AddBlobs`, `AddQueues`, and `AddTables` APIs.

**Comment by davidfowl** [comment]
I’m not sure this is a valuable breaking change to make.

**Comment by eerhardt** [comment]
> I’m not sure this is a valuable breaking change to make.

We could do it without a breaking change. We could add the new `.Get` methods and obsolete / hide the `.Add` methods.

**Comment by eerhardt** [comment]
The part that is confusing to me is that you aren't really "adding" a blobs service to a storage account. You get a blobs service (as well as queues and tables services) for every account. 

You are only "adding" the blobs service to the .NET Aspire model. And you only need a name for .NET Aspire. It doesn't show up in Azure at all.

**Comment by davidfowl** [comment]
> You are only "adding" the blobs service to the .NET Aspire model. And you only need a name for .NET Aspire. It doesn't show up in Azure at all.

Right but this is the an aspire API for Azure storage, not an azure storage API. Each of those services gets their own connection string. In theory, they don't need to be AddX, they could be properties with names derived from the storage resource.

The advantage of keeping the Add is that there's intent from the user that they want to use blobs/queues/tables. We could (and should) use that to inform role assignments.

**Comment by IEvangelist** [comment]
> > I’m not sure this is a valuable breaking change to make.
> 
> We could do it without a breaking change. We could add the new `.Get` methods and obsolete / hide the `.Add` methods.

Those are still technically considered breaking changes though, but that's that way I'd recommend doing it.

**Comment by davidfowl** [comment]
If we pivot this API away from the current model, do we still have names or are those resources get a name derived from storage? Customers also have to know the name because it needs to be referenced by the client app as a connection string.

**Comment by eerhardt** [comment]
> If we pivot this API away from the current model, do we still have names or are those resources get a name derived from storage? Customers also have to know the name because it needs to be referenced by the client app as a connection string.

I'm not sure. Maybe the name is derived from storage.

```C#
var storage = builder.AddAzureStorage("mystorage");

var blobs = storage.GetBlobs(); // implied name "mystorage-blobs"
var queues = storage.GetQueues(); // implied name "mystorage-queues"
var table = storage.GetTables(); // implied name "mystorage-tables"
```

Or you pass the name into the `.Get` method:

```C#
var storage = builder.AddAzureStorage("mystorage");

var blobs = storage.GetBlobs("blobs");
var queues = storage.GetQueues("queues");
var table = storage.GetTables("tables");
```

**Comment by eerhardt** [comment]
I agree this change may not have more value than it is worth.

**Comment by RussKie** [comment]
I agree that the current API are confusing - the me those look as though there are multiple "Blobs", "Queues" or "Tables".

In essence, those API return pointers to the Azure resource nodes (under the cover being represented by respective "service clients"):
![Image](https://github.com/user-attachments/assets/a15fb389-6226-462a-a38b-609b5a595d45)

So, whilst writing the following is totally possible, I can't think of a scenario where it'd make sense:
```cs
var storage = builder.AddAzureStorage("storage").RunAsEmulator(container =>
{
    container.WithDataBindMount();
});

var blobs1 = storage.AddBlobs("blobs1");
var blobs2 = storage.AddBlobs("blobs2");
```

I'm not sure renaming the API is something worth considering - it will be a breaking change (for a sake of a change), and it will diverge from the established naming convention (IIUIC) - "Add<Resource>" and "With<Options>".
Maybe, besides calling this explicitly in the docs, we could disallow multiple calls on the same storage? 
```cs
var blobs1 = storage.AddBlobs("blobs1");
var blobs2 = storage.AddBlobs("blobs2"); // <-- throws IOE?
```

**Comment by RussKie** [comment]
I propose to fix it in this way --> #9101

**Comment by RussKie** [comment]
There may still be a use-case for needing a service client (e.g., `BlobServiceClient`, `QueueServiceClient`, etc.) which are currently facilitated by `AddBlobs`, `AddQueues`, etc. API. And since there's now a desire to invest efforts in improving the API, I suggest considering renaming those API as well.

E.g.,:
```diff
- var blobs = storage.AddBlobs(..);
+ var blobs = storage.GetBlobServiceClient(..);

- var container = blobs.AddBlobContainer(...);
+ var container = storage.AddBlobContainer(...);

builder.AddProject<...>("api")
       .WithExternalHttpEndpoints()
       .WithReference(blobs).WaitFor(blobs)
       .WithReference(container).WaitFor(container);
```

**Comment by eerhardt** [comment]
> var blobs = storage.GetBlobServiceClient(..);

We shouldn't call this "BlobServiceClient" since that would be confusing with the client integration APIs - which are all suffixed with "Client".

**Comment by davidfowl** [comment]
I think we leave the existing APIs for this scenario. You still need a connection name and we already expose the endpoints as direct outputs.

**Comment by davidfowl** [comment]
I'd like us to fix this in 9.4 (add the new apis).

**Comment by sebastienros** [comment]
```diff
public static class AzureStorageExtensions
{

+   [Obsolete("Use AddBlobContainer on IResourceBuilder<AzureStorageResource> instead.")]
    public static IResourceBuilder<AzureBlobStorageContainerResource> AddBlobContainer(this IResourceBuilder<AzureBlobStorageResource> builder, [ResourceName] string name, string? blobContainerName = null)

+   public static IResourceBuilder<AzureBlobStorageContainerResource> AddBlobContainer(this IResourceBuilder<AzureStorageResource> builder, [ResourceName] string name, string? blobContainerName = null)

    public static IResourceBuilder<AzureQueueStorageResource> AddQueues(this IResourceBuilder<AzureStorageResource> builder, [ResourceName] string name)

+   public static IResourceBuilder<AzureQueueResource> AddQueue(this IResourceBuilder<AzureStorageResource> builder, [ResourceName] string name, string? queueName = null)

    public static IResourceBuilder<AzureTableStorageResource> AddTables(this IResourceBuilder<AzureStorageResource> builder, [ResourceName] string name)

+   public static IResourceBuilder<AzureTableResource> AddTable(this IResourceBuilder<AzureStorageResource> builder, [ResourceName] string name, string? queueName = null)
}

+ public class AzureQueueResource(string name, string queueName, AzureQueueStorageResource parent) { }

+ public class AzureTableResource(string name, string tableName, AzureTableStorageResource parent) { }
```

Notes:
- `AzureQueueResource` might be better named `AzureStorageQueueResource` even though it could be confusing with `AzureQueueStorageResource` (same for Table)
- There currently is `AddAzureBlobClient` which returns a `BlobServiceClient`. This method would make more sense returning a `BlobClient`. We'd still need a way to return a `BlobServiceClient`. Should we introduce a breaking change with `AddAzureBlobClient`? (same for Table and Queue).
- `AddQueue` and `AddTable` are currently non-existent, unlike `AddBlobContainer` which we are introducting at the top level.
- Corresponding CDK resources are created in the bicep files.

**Comment by eerhardt** [comment]
> AzureQueueResource might be better named AzureStorageQueueResource even though it could be confusing with AzureQueueStorageResource (same for Table)

Following the blobs naming, what do you think of `AzureQueueStorageQueueResource`?

> Should we introduce a breaking change with AddAzureBlobClient?

Do we have to right now?
